torch/csrc/autograd/autograd.cpp

```diff
@@ -0,0 +1,129 @@
+#include <torch/csrc/autograd/autograd.h>
+#include <torch/csrc/autograd/variable.h>
+
+#include <torch/csrc/autograd/edge.h>
+#include <torch/csrc/autograd/engine.h>
+#include <torch/csrc/autograd/function.h>
+
+
+namespace torch {
+namespace autograd {
+   variable_list run_backward(at::TensorList outputs,
+                              at::TensorList grad_outputs,
+                              bool keep_graph,
+                              bool create_graph,
+                              at::TensorList inputs) {
+    size_t num_tensors = outputs.size();
+    size_t num_gradients = grad_outputs.size();
+    TORCH_CHECK(num_tensors == num_gradients, "got %ld tensors and %ld "
+                "gradients", num_tensors, num_gradients);
+
+    edge_list roots;
+    roots.reserve(num_tensors);
+    variable_list grads;
+    grads.reserve(num_tensors);
+
+    for (size_t i = 0; i < num_tensors; i++) {
+        at::Tensor output = outputs[i];
+        TORCH_CHECK(output.is_variable(), "element %d of tensors "
+            "tuple is not a Tensor", i);
+        const Variable& variable = as_variable_ref(output);
+        auto gradient_edge = variable.gradient_edge();
+        TORCH_CHECK(gradient_edge.function,
+            "element %d of tensors does not require grad and does not have a grad_fn", i);
+        roots.push_back(std::move(gradient_edge));
+
+        at::Tensor grad = grad_outputs[i];
+        if (!grad.defined()) {
+            // if grad is undefined
+            if (output.requires_grad()) {
+                TORCH_CHECK(output.numel() == 1, "grad can be implicitly created only for scalar outputs");
+                // create the grad with ones tensor for output that requires grad
+                grads.push_back(at::ones_like(output));
+            }
+            // otherwise we don't put grad into the grads list
+        } else if (grad.is_variable()) {
+            // if grad is a typical variable
+            // grads.push_back(as_variable_ref(grad));
+            grads.push_back(as_variable_ref(grad));
+            std::cout<<"grad :" << grads[0] << std::endl;
+        } else {
+            AT_ERROR("element %d of gradients tuple is not a Tensor or None", i);
+        }
```

Comment:
IIUC part of the code here is based on `_make_grads()` in `torch/autograd/__init__.py` - since we are directly replicating all the logic in that file, it would be great to preserve the original function structure wrt. `backward()`/`grad()` calling `_make_grads()` first and then call `run_backward()`, which makes Python/C++ code parity maintenance easier in the future.

Reply:
Or can we just call the C++ implementation directly from Python?

Reply:
@ezyang that would make more sense to me on avoid maintaining two equivalent APIs, do you prefer me making it into a separate PR or combine things into this PR?

Reply:
Up to you, I generally like doing it in the same PR, mostly because it prevents us from forgetting to deduplicate later ;)

Reply:
Though, agreeing with Will, it's weird that the logic isn't replicated faithfully here

Reply:
> Or can we just call the C++ implementation directly from Python?

+1 here, if it's not too hard, we should just call the C++ implementation directly from Python.

Reply:
Yeah I will investigate how hard it is to direct call this API from python. For the reason that the logic isn't faithfully replicate here, it's because the current python API is making None objects first for `grad_outputs` and then `make_grads` construct another variable list with `ones_like` replacing some Nones, then in the THPEngine_run_backward we loop over the list again to get rid of Nones.  C++ API does not have None but only have undefined tensor, so if we fully replicate the logic we will constantly create those undefined tensors for those loops, which might not be efficient, we can just wrap the whole logic in `make_grads` to make it more clear.